pkg/authorization/reverse_permission_claims_authorizer.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	// claim is not accepted by the user that is importing it,
+	// hence the authorizer allows accessing the requested resource.
+	if a.claim.State != apisv1alpha1.ClaimAccepted {
+		return authorizer.DecisionAllow, fmt.Sprintf("claim is not accepted in API binding name=%q", a.binding.GetName()), nil
+	}
+
+	requestingAllResources := attr.GetName() == "" && attr.GetNamespace() == ""
+	resourceSelectorMatches := a.claim.All || requestingAllResources
+	for _, claimedResource := range a.claim.ResourceSelector {
+		var ns string
+		if claimedResource.Namespace != "" {
+			ns = attr.GetNamespace()
+		}
+		var name string
+		if claimedResource.Name != "" {
+			name = attr.GetName()
+		}
+		if claimedResource.Name == name && claimedResource.Namespace == ns {
+			resourceSelectorMatches = true
+			break
+		}
+	}
+
+	if !resourceSelectorMatches {
+		return authorizer.DecisionAllow, fmt.Sprintf("resource doesn't match any resource selector in API binding name=%q", a.binding.GetName()), nil
+	}
+
+	if len(a.claim.Verbs.RestrictTo) == 1 && a.claim.Verbs.RestrictTo[0] == "*" {
+		return authorizer.DecisionAllow, fmt.Sprintf("permission claim restricts to \"*\" in API binding name=%q", a.binding.GetName()), nil
+	}
+
+	for _, verb := range a.claim.Verbs.RestrictTo {
+		if attr.GetVerb() == verb {
+			return authorizer.DecisionAllow, fmt.Sprintf("verb is restricted to in permission claim in API binding name=%q", a.binding.GetName()), nil
+		}
+	}
+
+	return authorizer.DecisionDeny, fmt.Sprintf("verb is not restricted to in API binding name=%q", a.binding.GetName()), nil
+}
+
+func getAPIBindingsMatchingClaimsForAttributes(apiBindingsLister apisv1alpha1listers.APIBindingClusterLister, attr authorizer.Attributes, clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+	bindings, err := apiBindingsLister.Cluster(clusterName).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	var result []*apisv1alpha1.APIBinding
+	for _, apiBinding := range bindings {
+		for _, claim := range apiBinding.Spec.PermissionClaims {
+			if claim.State != apisv1alpha1.ClaimAccepted {
+				continue
+			}
+			if claim.Group == attr.GetAPIGroup() && claim.Resource == attr.GetResource() {
+				result = append(result, apiBinding)
+			}
+		}
+	}
+	return result, nil
+}
+
+type IntersectionAuthorizer []authorizer.Authorizer
+
+func (auth IntersectionAuthorizer) Authorize(ctx context.Context, a authorizer.Attributes) (authorizer.Decision, string, error) {
+	var (
+		errlist    []error
+		reasonlist []string
+	)
+
+	for _, curr := range auth {
+		decision, reason, err := curr.Authorize(ctx, a)
+
+		if err != nil {
+			errlist = append(errlist, err)
+		}
+		if len(reason) != 0 {
+			reasonlist = append(reasonlist, reason)
+		}
+		switch decision {
+		case authorizer.DecisionDeny:
+			return decision, reason, err
+		case authorizer.DecisionNoOpinion:
+			// continue to the next authorizer
+		}
```

Comment:
same as above, this is from upstream's union authorizer. I suggest to leave it here so it can be easier compared to the union authorizer (although upstream can change any time).